### ISSUE TITLE
Add `requireParamTypes` option to the `valid-jsdoc` rule

### ISSUE THREAD

**Issue by kripod** [eslint/eslint#6753]
While using [babel-plugin-typecheck](https://github.com/codemix/babel-plugin-typecheck), I realized that specifying parameter types is redundant, as they can be retrieved from the [Flow](https://github.com/facebook/flow) syntax.

See [TypeDoc's DocComments](http://typedoc.io/guides/doccomments.html) for further reference about the accepted documentation format with `requireParamTypes` set to `false`. The default value of `requireParamTypes` should be `true`.

**Comment by eslintbot** [comment]
Hi @kripod, thanks for the issue. It looks like there's not enough information for us to know how to help you. If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

If it's something else, please just provide as much additional information as possible. Thanks!

**Comment by kripod** [comment]
# enforce valid JSDoc comments (valid-jsdoc)
## Options

* `"requireParamTypes": false` allows missing type in parameter tags

### requireParamTypes

Example of additional **correct** code for this rule with the `"requireReturn": false` option:

```js
/*eslint valid-jsdoc: ["error", { "requireParamTypes": false }]*/

/**
 * Add two numbers.
 * @param num1 The first number.
 * @param num2 The second number.
 * @returns {number} The sum of the two numbers.
 */
function add(num1, num2) {
    return num1 + num2;
}
```

Example of a use case with [Flow](https://flowtype.org/) type annotations:

```js
/*eslint valid-jsdoc: ["error", { "requireParamTypes": false, "requireReturnType": false }]*/

/**
 * Add two numbers.
 * @param num1 The first number.
 * @param num2 The second number.
 * @returns The sum of the two numbers.
 */
function add(num1: number, num2: number): number {
    return num1 + num2;
}
```

---

The rule should be in the core because it extends the functionality of an existent core rule (`valid-jsdoc`). I'm not willing to create the rule myself.

**Comment by vitorbal** [comment]
Hi @kripod, thanks for your submission. Please take a moment to read over our guide on proposing rule changes: http://eslint.org/docs/developer-guide/contributing/new-rules

Specifically, we do require proposers to implement their suggestions if accepted, so your willingness to implement it is a key decision point.

Currently, the team is very focused on improving compatibility between JSCS and ESLint, so if you are not willing to work on a PR for this, I'm afraid there is a low chance this will end up getting accepted.

**Comment by ilyavolodin** [comment]
I'm slightly confused, as far as I understand, jsDoc exists for documentation purposes, and Flow for pre-compile type validation. I'm not sure I see why having Flow should affect not having good documentation. @kripod Could you explain please?

**Comment by kripod** [comment]
Good documentation could be generated by inheriting expected types from Flow syntax, and combining them with descriptions specified using JSDoc comments. Similar to how TSDoc and C#'s built-in XML documentation provider work, JSDoc should not require parameter types to be provided explicitly whether they are already clearly defined by Flow syntax.

**Comment by ilyavolodin** [comment]
@kripod I'm not sure if you are explaining ideal situation or what is happening right now. Are there any tools out there that would generate documentation by combining jsDoc and Flow?
And in general, ESLint doesn't naively support Flow. It's only supported through `babel-eslint`. So we would need much more generic scenario to consider adding an option to disable typing for jsDoc in the core. Otherwise, this might be a good request to add to [`eslint-plugin-flow`](https://www.npmjs.com/package/eslint-plugin-flowtype)

**Comment by kripod** [comment]
If I recall correctly, [documentation.js](https://github.com/documentationjs/documentation) supports Flow type annotations. I'll submit an issue for [eslint-plugin-flowtype](https://github.com/gajus/eslint-plugin-flowtype) based on your suggestion, but I'm leaving this issue open for further discussion. Thank you for your assistance!

**Comment by vitorbal** [comment]
I think we can continue the discussion on the [eslint-plugin-flowtype issue you opened](https://github.com/gajus/eslint-plugin-flowtype/issues/37), so I will close this. We can always re-open if anyone thinks there's still something to be discussed here.

**Comment by mathieumg** [comment]
@vitorbal 
Is there a way for `eslint-plugin-flowtype` to extend `valid-jsdoc` without copy-pasting it? I thought there wasn't. I see there is already `requireReturnType` and `requireParamDescription` supported by `valid-jsdoc`, why couldn't `requireParamType` also be, for the sake of consistency? Flow was discussed as one of the reasons why someone would not want types. But there are others; they could be using another typing engine or they could simply want descriptions and not types in their docblocks.

Are you open to the idea of re-opening this issue? I could take a stab at the implementation. Basically, it just brings more consistency to the options offered by the rule and would be off by default, obviously.

Thanks!
